azure-iot-device/azure/iot/device/iothub/sync_clients.py

```diff
@@ -173,9 +167,30 @@ def disconnect(self):
         """
         logger.info("Disconnecting from Hub...")
 
+        logger.debug("Executing initial disconnect")
         callback = EventedCallback()
         self._mqtt_pipeline.disconnect(callback=callback)
         handle_result(callback)
+        logger.debug("Successfully executed initial disconnect")
+
+        # Note that in the process of stopping the handlers and resolving pending calls
+        # a user-supplied handler may cause a reconnection to occur
+        logger.debug("Stopping handlers...")
+        self._handler_manager.stop()
+        logger.debug("Successfully stopped handlers")
+
+        # Disconnect again to ensure disconnection has ocurred due to the issue mentioned above
+        logger.debug("Executing secondary disconnect...")
+        callback = EventedCallback()
+        self._mqtt_pipeline.disconnect(callback=callback)
+        handle_result(callback)
+        logger.debug("Successfully executed secondary disconnect")
+
+        # It's also possible that in the (very short) time between stopping the handlers and
+        # the second disconnect, additional items were received (e.g. C2D Message)
+        # Currently, this isn't really possible to accurately check due to a race condition.
+        # It has always been true of this client, even before handlers.
+        # TODO: Fix that
```

Comment:
This actually may not be completely fixable on our end - may need to wait for MQTT 5 to completely resolve this issue (although there are things we can do to mitigate it in the meantime)